lib/rules/no-misleading-character-class.js

```diff
@@ -195,6 +195,55 @@ module.exports = {
         const sourceCode = context.sourceCode;
         const parser = new RegExpParser();
 
+        /**
+         * Determines what offset to start a granular report, if it's straightforward to do so.
+         * @param {string} char Individual character being reported on.
+         * @param {Node} node Parent string node to report within.
+         * @see https://github.com/eslint/eslint/pull/17515#issuecomment-1707697186s
+         * @returns {Object | null} Report offset, or null to report on the whole node.
+         */
+        function getReportStartOffset(char, node) {
+            if (node.type !== "Literal" && (node.type !== "TemplateLiteral" || node.expressions.length)) {
+                return null;
+            }
+
+            const offsetStart = node.raw.slice(0, char.start + char.raw.length + 2).indexOf(char.raw, char.start - 2);
+
+            return offsetStart === -1 ? null : offsetStart;
+        }
+
+        /**
+         * Generates a granular loc for context.report.
+         * @param {string} char Individual character being reported on.
+         * @param {string} kind What kind of character report is being made.
+         * @param {Node} node Parent string node to report within.
+         * @param {number} reportStartOffset Starting offset to report within the node.
+         * @returns {Object} Granular loc for context.report.
+         * @see https://github.com/eslint/eslint/pull/17515#issuecomment-1707697186s
+         */
+        function generateReportLocation(char, kind, node, reportStartOffset) {
+            const reportEndOffset = reportStartOffset + char.raw.length;
+            let reportStartOffsetAdjusted = reportStartOffset;
+
+            if (!char.raw.startsWith("\\u")) {
+                reportStartOffsetAdjusted -=
+                    ["combiningClass", "surrogatePairWithoutUFlag"].includes(kind)
+                        ? 1
+                        : 2;
+            }
```

Comment:
This is a bit difficult to visually parse.

Because this array never changes, I'd suggest moving it out of this function into a constant. Then, you could just calculate the adjustment and add it to reportStartOffset at the end of this function.

```js
adjustment = kinds.includes(kind) ? 1 : 2;
```

Then on line 242:

```js
column: node.loc.start.column + reportStartOffset - adjustment
```

Reply:
Makes sense, thanks! Since this only happens if `char.raw` doesn't start with `"\\u"`, in c370130e2 I went with just the extracting into a constant. I suppose an alternative would be to default `adjustment` to `0`, but that felt like more lines to me. WDYT?